src/DS/metrics_extraction.py

```diff
@@ -0,0 +1,300 @@
+            "has_teardown": has_teardown,
+            "complexity": complexity,
+        }
+    )
+
+    return result
+
+
+def analyse_code_file(file_path, max_complexity=10):
+    """
+    Analyses a Python code file to extract code metrics such as cyclomatic
+    complexity, lines of code, and the number of functions.
+
+    Args:
+        file_path (str): The path to the code file.
+        max_complexity (int): The maximum allowed complexity for a function.
+
+    Returns:
+        dict: A dictionary containing the code metrics.
+    """
+
+    logger.info(f"Starting analysis of code file: {file_path}")
+
+    result = {
+        "cyclomatic_complexity": 0,
+        "lines_of_code": 0,
+        "num_functions": 0,
+    }
+
+    parsed_file = read_and_parse_file(file_path)
+    if parsed_file is None:
+        logger.warning(f"Analysis skipped for file: {file_path}")
+        return result
+
+    tree, content = parsed_file
+
+    num_functions = 0
+    lines_of_code = len(content.splitlines())
+    cyclomatic_complexity = 0
+
+    visitor = PathGraphingAstVisitor()
+    visitor.preorder(tree, visitor)
+
+    for graph in visitor.graphs.values():
+        "Each graph in visitor.graphs.values() represents the control flow " "graph (CFG) of a single function, which is used to compute its " "cyclomatic complexity."
+        "'PathGraphingAstVisitor' processes the Abstract Syntax Tree (AST) of" "the code and generates a control flow graph for each function it" "encounters, storing these graphs in 'visitor.graphs'. Each entry" "in 'visitor.graphs' corresponds to one function, so iterating over" "it and counting with 'num_functions += 1' accurately captures the" "function count."
+
+        num_functions += 1
+        complexity = graph.complexity()
+        cyclomatic_complexity += complexity
+        logger.debug(f"Function {graph.name} has complexity {complexity}")
+
+    result.update(
+        {
+            "cyclomatic_complexity": cyclomatic_complexity,
+            "lines_of_code": lines_of_code,
+            "num_functions": num_functions,
+        }
+    )
+
+    return result
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    input_file = args.input
+    output_file = args.output
+
+    working_directory = get_working_directory_or_git_root()
+    logger.info(f"Working directory: {working_directory}")
+
+    df = pd.read_csv(input_file)
+    df_language_python = df[df["guessed_language"] == "Python"]
+
+    final_df_path = output_file
+    logger.info("Apply the analysis to each test file")
+
+    # Initialise or load the final Dataframe
+    if Path(final_df_path).exists():
+        final_df = pd.read_csv(final_df_path)
+        logger.info(f"Loaded existing final Dataframe from {final_df_path}")
+
+    else:
+        final_df = pd.DataFrame()
+
+    batch_results = []
+
+    for idx, row in df_language_python.iterrows():
+        file_path = row["file_path"]
+        test_file_analysis = analyse_test_file(file_path)
```

Comment:
Is every python file processed twice, once to gather statistics about tests they contain (if any) and once to gather complexity related statistics? If so, this approach increases the potential of finding testing related code regardless of the filename, however it also increases the overall file processing. I don't have a strong opinion on which approach is more useful so we can try this approach and see what the performance is and the results indicate.

Reply:
Yes this is correct. Each Python file is processed twice. I was thinking there might be some relations between the two. For instance, the relation between number of test cases(' num_test_cases') and number of functions(`num_functions`).